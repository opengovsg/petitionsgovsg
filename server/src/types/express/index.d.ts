declare namespace Express {
  interface User {
    id: string
    type: string
    fullname: string
  }
}
