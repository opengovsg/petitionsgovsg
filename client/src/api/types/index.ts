export * from './post'
export * from './signature'
export * from './addressee'
export * from './auth'
