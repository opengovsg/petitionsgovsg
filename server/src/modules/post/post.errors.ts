import { StatusCodes } from 'http-status-codes'
import { ApplicationError } from '../core/core.errors'

export class MissingPublicPostError extends ApplicationError {
  constructor(
    message = 'No public post with this id',
    statusCode = StatusCodes.NOT_FOUND,
  ) {
    super(message, statusCode)
  }
}

export class PostUpdateError extends ApplicationError {
  constructor(
    message = 'Post update failed',
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
  ) {
    super(message, statusCode)
  }
}
