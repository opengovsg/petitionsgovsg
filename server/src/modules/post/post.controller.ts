import { validationResult } from 'express-validator'
import { StatusCodes } from 'http-status-codes'
import { Post } from '~shared/types/base'
import { createLogger } from '../../bootstrap/logging'
import { Message } from '../../types/message-type'
import { ControllerHandler } from '../../types/response-handler'
import { SortType } from '../../types/sort-type'
import { AuthService } from '../auth/auth.service'
import { PostService, PostWithAddresseeAndSignatures } from './post.service'
import { hashData, generateSalt } from '../../util/hash'

const logger = createLogger(module)

export class PostController {
  private authService: Public<AuthService>
  private postService: Public<PostService>

  constructor({
    authService,
    postService,
  }: {
    authService: Public<AuthService>
    postService: Public<PostService>
  }) {
    this.authService = authService
    this.postService = postService
  }

  /**
   * Lists all post
   * @query sort Sort by popularity or recent
   * @query size Number of posts to return
   * @query page If size is given, specify which page to return
   * @return 200 with posts and totalItem for pagination
   * @return 500 when database error occurs
   */
  listPosts: ControllerHandler<
    undefined,
    { posts: Post[]; totalItems: number } | Message,
    undefined,
    {
      page?: number
      size?: number
      sort?: SortType
    }
  > = async (req, res) => {
    const { page, size, sort = SortType.Top } = req.query
    try {
      const data = await this.postService.listPosts({
        sort: sort as SortType,
        page: page,
        size: size,
      })
      return res.status(StatusCodes.OK).json(data)
    } catch (error) {
      if (error instanceof Error) {
        logger.error({
          message: 'Error while listing posts',
          meta: {
            function: 'listPosts',
          },
          error,
        })
        return res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Server Error' })
      }
    }
  }

  /**
   * Get a single post and all the users associated with it
   * @param postId Id of the post
   * @query relatedPosts if true, return related posts
   * @return 200 with post
   * @return 403 if user does not have permission to access post
   * @return 500 for database error
   */
  getSinglePost: ControllerHandler<
    { id: number },
    PostWithAddresseeAndSignatures | Message,
    undefined,
    { relatedPosts?: number }
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    let post
    try {
      post = await this.postService.getSinglePost(req.params.id)
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server Error' })
    }

    try {
      await this.authService.verifyUserCanViewPost(
        post,
        req.user?.id.toString() ?? '',
      )
    } catch (error) {
      logger.error({
        message: 'Error while retrieving single post',
        meta: {
          function: 'getSinglePost',
        },
        error,
      })
      return res
        .status(StatusCodes.FORBIDDEN)
        .json({ message: 'User does not have permission to access this post' })
    }

    return res.status(StatusCodes.OK).json(post)
  }

  /**
   * Create a new post
   * @body title title of post
   * @body summary summary of post
   * @body reason reason of post
   * @body request request of post
   * @body references references of post
   * @return 200 if post is created
   * @return 400 if title and description is too short or long
   * @return 401 if user is not signed in
   * @return 500 if database error
   */
  createPost: ControllerHandler<
    undefined,
    { data: number } | Message,
    {
      title: string
      summary: string | null
      reason: string
      request: string
      references: string | null
      fullname: string
      addresseeId: number
      profile: string | null
      email: string
      salt: string
    },
    undefined
  > = async (req, res) => {
    const errors = validationResult(req)
    if (!errors.isEmpty()) {
      return res.status(StatusCodes.BAD_REQUEST).json(errors.array()[0].msg)
    }
    if (!req.user) {
      return res
        .status(StatusCodes.UNAUTHORIZED)
        .json({ message: 'User not signed in' })
    }

    try {
      if (!req.user) {
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'User not signed in' })
      }

      const salt = await generateSalt()
      const hashedUserSgid = await hashData(req.user.id, salt)
      const data = await this.postService.createPost({
        title: req.body.title,
        summary: req.body.summary,
        reason: req.body.reason,
        request: req.body.request,
        hashedUserSgid: hashedUserSgid,
        references: req.body.references,
        fullname: req.body.fullname,
        addresseeId: req.body.addresseeId,
        profile: req.body.profile,
        email: req.body.email,
        salt: salt,
      })

      return res.status(StatusCodes.OK).json({ data: data })
    } catch (error) {
      logger.error({
        message: 'Error while creating post',
        meta: {
          function: 'addPost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }

  /**
   * Delete a post
   * @param id Post to be deleted
   * @return 200 if successful
   * @return 401 if user is not logged in
   * @return 403 if user does not have permission to delete post
   * @return 500 if database error
   */
  deletePost: ControllerHandler<{ id: string }, Message> = async (req, res) => {
    const postId = Number(req.params.id)
    try {
      const userId = req.user?.id
      if (!userId) {
        logger.error({
          message: 'UserId is undefined after authenticated',
          meta: {
            function: 'deletePost',
          },
        })
        return res
          .status(StatusCodes.UNAUTHORIZED)
          .json({ message: 'You must be logged in to delete posts.' })
      }
      // TODO
      // const hasPermission = await this.authService.hasPermissionToEditPost(
      //   userId,
      //   postId,
      // )
      // if (!hasPermission) {
      //   return res
      //     .status(StatusCodes.FORBIDDEN)
      //     .json({ message: 'You do not have permission to delete this post.' })
      // }
      await this.postService.deletePost(postId)
      return res.status(StatusCodes.OK).send({ message: 'OK' })
    } catch (error) {
      logger.error({
        message: 'Error while deleting post',
        meta: {
          function: 'deletePost',
        },
        error,
      })
      return res
        .status(StatusCodes.INTERNAL_SERVER_ERROR)
        .json({ message: 'Server error' })
    }
  }
}
