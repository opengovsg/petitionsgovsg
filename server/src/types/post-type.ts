import { Post } from '~shared/types/base'

export type PostEditType = {
  id: number
  userid: number
  summary: string
  title: string
}

export type UpdatePostRequestDto = Pick<Post, 'title' | 'summary'>
