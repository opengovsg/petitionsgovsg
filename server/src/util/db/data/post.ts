import { PostStatus } from '~shared/types/base'

export const mockPost = {
  id: '395ac6ec-dea2-4f78-8dc4-1234cafaff87',
  title: 'Basic petition without optional items',
  reason: 'reason for',
  request: 'request',
  status: PostStatus.Open,
  fullname: 'John Doe',
  salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
  hashedUserSgid: 'wxyz',
  email: 'John@email.com',
  addresseeId: 1,
  summary: null,
  references: null,
  profile: null,
  createdAt: '2022-01-01T00:00:00+00:00',
  updatedAt: '2022-01-01T00:00:00+00:00',
}

export const mockPosts = [
  {
    id: 'af7b6aa9-7c6e-4ea5-b536-4e55bc506847',
    title: 'Basic petition without optional items',
    reason: 'reason for',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tW.',
    hashedUserSgid:
      '$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW',
    email: 'John@email.com',
    addresseeId: 1,
    summary: null,
    references: null,
    profile: null,
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
  {
    id: 'aaaaaaaa-dea2-4f78-8dc4-5542cafaff87',
    title: 'Petition with summary, references, profile',
    reason: 'reason',
    request: 'request',
    status: PostStatus.Open,
    fullname: 'John Doe',
    salt: '$2b$10$72TKINiR4.ZMaJZmuN5tX.',
    hashedUserSgid: 'wxyz',
    email: 'John@email.com',
    addresseeId: 1,
    summary: 'petition summary',
    references: 'wikipedia.com',
    profile: 'OGP',
    createdAt: '2022-01-01T00:00:00+00:00',
    updatedAt: '2022-01-01T00:00:00+00:00',
  },
]
