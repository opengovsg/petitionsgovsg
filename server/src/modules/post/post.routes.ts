import { AuthMiddleware } from '../auth/auth.middleware'
import express from 'express'
import { check, param } from 'express-validator'
import { PostController } from './post.controller'

export const routePosts = ({
  controller,
  authMiddleware,
}: {
  controller: PostController
  authMiddleware: Public<AuthMiddleware>
}): express.Router => {
  const router = express.Router()

  /**
   * Lists all post
   * @route   GET /api/posts
   * @returns 200 with posts
   * @returns 500 when database error occurs
   * @access  Public
   */
  router.get('/', controller.listPosts)

  /**
   * Get a single post and all users associated with it
   * @route  GET /api/posts/:id
   * @return 200 with post
   * @return 403 if user does not have permission to access post
   * @return 500 for database error
   * @access Public
   */
  router.get('/:id', [param('id').isInt().toInt()], controller.getSinglePost)

  /**
   * Create a new post
   * @route  POST /api/posts/
   * @return 200 if post is created
   * @return 400 if title and description is too short or long
   * @return 401 if user is not signed in
   * @return 500 if database error
   * @access Private
   */
  router.post(
    '/',
    [
      authMiddleware.authenticate,
      check(
        'title',
        'Enter a title with minimum of 15 characters and maximum 150 characters',
      ).isLength({
        min: 15,
        max: 150,
      }),
      check('email', 'Enter an email with a minimum of 1 characters').isLength({
        min: 1,
      }),
    ],
    controller.createPost,
  )

  /**
   * Delete a post
   * @route  DELETE /api/posts/:id
   * @return 200 if successful
   * @return 401 if user is not logged in
   * @return 403 if user does not have permission to delete post
   * @return 500 if database error
   * @access Private
   */
  router.delete(
    '/:id([0-9]+$)',
    [authMiddleware.authenticate],
    controller.deletePost,
  )

  return router
}
