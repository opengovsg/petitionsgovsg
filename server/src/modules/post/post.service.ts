import type { Sequelize as SequelizeType } from 'sequelize'
import Sequelize, { Model, Op, OrderItem, ProjectionAlias } from 'sequelize'
import {
  ListPostsDto,
  PostWithAddresseeAndSignatures,
  UpdatePostReqDto,
} from '~shared/types/api'
import { Addressee, Post, PostStatus, Signature } from '~shared/types/base'
import { ModelDef } from '@/types/sequelize'
import { SortType } from '~shared/types/base/post'
import {
  AddresseeDoesNotExistError,
  MissingPublicPostError,
  PostUpdateError,
} from './post.errors'

export class PostService {
  private Signature: ModelDef<Signature>
  private Post: ModelDef<Post>
  private Addressee: ModelDef<Addressee>
  private sequelize: SequelizeType

  constructor({
    Signature,
    Post,
    Addressee,
    sequelize,
  }: {
    Signature: ModelDef<Signature>
    Post: ModelDef<Post>
    Addressee: ModelDef<Addressee>
    sequelize: SequelizeType
  }) {
    this.Signature = Signature
    this.Post = Post
    this.Addressee = Addressee
    this.sequelize = sequelize
  }

  private signatureCountLiteral: ProjectionAlias = [
    Sequelize.literal(`(
      SELECT COUNT(*)
      FROM signatures AS signature
      WHERE
        signature.postId = post.id
    )`),
    'signatureCount',
  ]

  private sortFunction = (sortType: SortType): OrderItem => {
    const sortDict = {
      [SortType.Newest]: ['createdAt', 'DESC'],
      [SortType.Oldest]: ['createdAt', 'ASC'],
      [SortType.MostSignatures]: [Sequelize.literal('signatureCount'), 'DESC'],
      [SortType.LeastSignatures]: [Sequelize.literal('signatureCount'), 'ASC'],
    }

    return sortDict[sortType] as OrderItem
  }
  /**
   * Return the paginated posts
   * @param posts post array to be paginated
   * @param page If size is given, specify which page to return
   * @param size Number of posts to return
   * @returns paginated post array and total items in the original array
   */
  private getPaginatedPosts = (
    posts: PostWithAddresseeAndSignatures[],
    page?: number,
    size?: number,
  ): {
    posts: PostWithAddresseeAndSignatures[]
    totalItems: number
  } => {
    const totalItems = posts.length
    let returnPosts = posts
    if (size) {
      const offset = page ? (page - 1) * size : 0
      returnPosts = returnPosts.slice(offset, Number(offset) + Number(size))
    }
    return {
      posts: returnPosts,
      totalItems: totalItems,
    }
  }

  /**
   * Lists all post
   * @param sort Sort by popularity or recent
   * @param size Number of posts to return
   * @param page If size is given, specify which page to return
   */
  listPosts = async ({
    sort,
    page,
    size,
  }: {
    sort: SortType
    page?: number
    size?: number
  }): Promise<ListPostsDto> => {
    const whereobj = {
      status: PostStatus.Open,
    }

    const orderarray = this.sortFunction(sort)

    const posts = (await this.Post.findAll({
      where: whereobj,
      order: [orderarray],
      include: [
        this.Signature,
        {
          model: this.Addressee,
          required: false,
          attributes: ['name', 'shortName'],
        },
      ],
      attributes: [
        'id',
        'createdAt',
        'updatedAt',
        'status',
        'title',
        'summary',
        'reason',
        'request',
        'hashedUserSgid',
        'references',
        'fullname',
        'addresseeId',
        'profile',
        'email',
        this.signatureCountLiteral,
      ],
    })) as (Model & PostWithAddresseeAndSignatures)[]

    if (!posts) {
      return { posts: [], totalItems: 0 }
    } else {
      return this.getPaginatedPosts(posts, page, size)
    }
  }

  /**
   * Get a single post and all users associated with it
   * @param postId Id of the post
   */
  getSinglePost = async (
    postId: number,
  ): Promise<PostWithAddresseeAndSignatures> => {
    const post = (await this.Post.findOne({
      where: {
        [Op.or]: [{ status: PostStatus.Open }, { status: PostStatus.Draft }],
        id: postId,
      },
      include: [
        this.Signature,
        {
          model: this.Addressee,
          required: false,
          attributes: ['name', 'shortName'],
        },
      ],
      attributes: [
        'createdAt',
        'updatedAt',
        'status',
        'title',
        'summary',
        'reason',
        'request',
        'hashedUserSgid',
        'references',
        'salt',
        'fullname',
        'addresseeId',
        'profile',
        'email',
        this.signatureCountLiteral,
      ],
    })) as Model & PostWithAddresseeAndSignatures
    if (!post) {
      throw new MissingPublicPostError()
    } else {
      return post
    }
  }

  /**
   * Create a new post
   * @param newPost Post to be created
   * @returns Id of the new post if it is successfully created
   */
  createPost = async (newPost: {
    title: string
    summary: string | null
    reason: string
    request: string
    hashedUserSgid: string
    references: string | null
    fullname: string
    addresseeId: number
    profile: string | null
    email: string
    salt: string
  }): Promise<number> => {
    // Check if addresseeId is valid
    const addresseeExists = await this.Addressee.findByPk(newPost.addresseeId)
    if (!addresseeExists) {
      throw new AddresseeDoesNotExistError()
    }
    const postId = await this.sequelize.transaction(async (transaction) => {
      const post = await this.Post.create(
        { ...newPost, status: PostStatus.Draft },
        { transaction },
      )
      return post.id
    })
    return postId
  }

  /**
   * Delete a post by setting the post to archive
   * @param id Post to be deleted
   * @returns void if successful
   */
  deletePost = async (id: number): Promise<void> => {
    try {
      await this.sequelize.transaction(async (transaction) => {
        const [dbUpdate] = await this.Post.update(
          { status: PostStatus.Closed },
          { where: { id: id }, transaction },
        )
        if (!dbUpdate) {
          throw new PostUpdateError()
        }
      })
    } catch (error) {
      throw error
    }
  }

  /**
   * Update a post
   * @param id Post to be updated
   * @returns boolean Indicate if update was successful
   */
  updatePost = async ({
    id,
    title,
    summary,
    reason,
    request,
    references,
    addresseeId,
    profile,
    email,
  }: UpdatePostReqDto & { id: number }): Promise<boolean> => {
    // Check if addresseeId is valid
    const addresseeExists = await this.Addressee.findByPk(addresseeId)
    if (!addresseeExists) {
      throw new AddresseeDoesNotExistError()
    }

    const updated = await this.sequelize.transaction(async (transaction) => {
      const dbUpdate = await this.Post.update(
        {
          title,
          summary,
          reason,
          request,
          references,
          addresseeId,
          profile,
          email,
        },
        {
          where: { id: id },
          transaction,
        },
      )
      return dbUpdate
    })
    return !!updated
  }

  /**
   * Change post status to Open
   * @param id Post to be published
   * @returns void if successful
   */
  publishPost = async (id: number): Promise<void> => {
    await this.sequelize.transaction(async (transaction) => {
      const [dbUpdate] = await this.Post.update(
        { status: PostStatus.Open },
        { where: { id: id }, transaction },
      )
      if (!dbUpdate) {
        throw new PostUpdateError()
      }
    })
  }
}
