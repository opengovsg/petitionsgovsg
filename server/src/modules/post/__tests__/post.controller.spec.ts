import express from 'express'
import { StatusCodes } from 'http-status-codes'
import supertest from 'supertest'
import { UserAuthType } from '~shared/types/api'
import { ControllerHandler } from '../../../types/response-handler'
import { PostController } from '../post.controller'

describe('PostController', () => {
  const path = '/posts'
  const authService = {
    hasPermissionToEditPost: jest.fn(),
    verifyUserCanViewPost: jest.fn(),
  }
  const postService = {
    createPost: jest.fn(),
    deletePost: jest.fn(),
    getSinglePost: jest.fn(),
    listPosts: jest.fn(),
  }
  const userService = {
    loadUser: jest.fn(),
    loadUserBySgid: jest.fn(),
  }
  const controller = new PostController({
    authService,
    postService,
    userService,
  })

  // Set up auth middleware to inject user
  let user: Express.User | undefined = { id: 1, type: UserAuthType.Public }
  const middleware: ControllerHandler = (req, res, next) => {
    req.user = user
    next()
  }

  beforeEach(() => {
    user = { id: 1, type: UserAuthType.Public }
  })
  afterEach(async () => {
    jest.clearAllMocks()
  })

  describe('listPosts', () => {
    beforeEach(() => {
      postService.listPosts.mockReset()
    })
    it('should return 200 on successful data retrieval', async () => {
      // Arrange
      const data = { rows: ['1', '2'], totalItems: 5 }
      postService.listPosts.mockReturnValue(data)

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.get(path, controller.listPosts)
      const request = supertest(app)

      // Act
      const response = await request.get(path)

      // Assert
      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual(data)
    })

    it('should return 500 on any other errors', async () => {
      // Arrange
      const error = new Error('Database error')
      postService.listPosts.mockImplementation(() => {
        throw error
      })

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.get(path, controller.listPosts)
      const request = supertest(app)

      // Act
      const response = await request.get(path)

      // Assert
      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({ message: 'Server Error' })
    })
  })

  describe('createPost', () => {
    const userId = 21
    const postId = 13

    beforeEach(() => {
      userService.loadUser.mockResolvedValue({ userId })
      postService.createPost.mockResolvedValue(postId)
    })
    it('returns 401 on no user', async () => {
      user = undefined

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.post(path, controller.createPost)
      const request = supertest(app)

      const response = await request.post(path)

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toStrictEqual({
        message: 'User not signed in',
      })
      expect(postService.createPost).not.toHaveBeenCalled()
    })
    it('returns 500 on userService problem', async () => {
      userService.loadUser.mockRejectedValue(new Error('bad user'))

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.post(path, controller.createPost)
      const request = supertest(app)

      const response = await request.post(path)

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({
        message: 'Server error',
      })
      expect(postService.createPost).not.toHaveBeenCalled()
    })
    it('returns 500 on postService problem', async () => {
      const body = { title: 'Title', summary: 'Summary' }

      postService.createPost.mockRejectedValue(new Error('bad post'))

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.post(path, controller.createPost)
      const request = supertest(app)

      const response = await request.post(path).send(body)

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({
        message: 'Server error',
      })
      expect(postService.createPost).toHaveBeenCalledWith({
        ...body,
        userId: user?.id,
      })
    })
    it('returns 200 on successful creation', async () => {
      const body = { title: 'Title', summary: 'Summary' }

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.post(path, controller.createPost)
      const request = supertest(app)

      const response = await request.post(path).send(body)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual({ data: postId })
      expect(postService.createPost).toHaveBeenCalledWith({
        ...body,
        userId: user?.id,
      })
    })
  })

  describe('deletePost', () => {
    const postId = 1

    beforeEach(() => {
      authService.hasPermissionToEditPost.mockReset()
      postService.deletePost.mockReset()
    })

    it('returns 200 on successful deletion', async () => {
      authService.hasPermissionToEditPost.mockResolvedValue(true)
      postService.deletePost.mockImplementation(() => Promise.resolve())

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.delete(path + '/:id([0-9]+$)', controller.deletePost)
      const request = supertest(app)

      const response = await request.delete(path + `/${postId}`)

      expect(response.status).toEqual(StatusCodes.OK)
      expect(response.body).toStrictEqual({
        message: 'OK',
      })
      expect(postService.deletePost).toHaveBeenCalledWith(postId)
    })

    it('returns 401 on no user', async () => {
      user = undefined

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.delete(path + '/:id([0-9]+$)', controller.deletePost)
      const request = supertest(app)

      const response = await request.delete(path + `/${postId}`)

      expect(response.status).toEqual(StatusCodes.UNAUTHORIZED)
      expect(response.body).toStrictEqual({
        message: 'You must be logged in to delete posts.',
      })
      expect(postService.deletePost).not.toHaveBeenCalled()
    })

    it('returns 403 if user does not have permission to delete post', async () => {
      authService.hasPermissionToEditPost.mockResolvedValue(false)

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.delete(path + '/:id([0-9]+$)', controller.deletePost)
      const request = supertest(app)

      const response = await request.delete(path + `/${postId}`)

      expect(response.status).toEqual(StatusCodes.FORBIDDEN)
      expect(response.body).toStrictEqual({
        message: 'You do not have permission to delete this post.',
      })
      expect(postService.deletePost).not.toHaveBeenCalled()
    })

    it('returns 500 on database error', async () => {
      authService.hasPermissionToEditPost.mockRejectedValue(
        new Error('Error while deleting post'),
      )

      const app = express()
      app.use(express.json())
      app.use(middleware)
      app.delete(path + '/:id([0-9]+$)', controller.deletePost)
      const request = supertest(app)

      const response = await request.delete(path + `/${postId}`)

      expect(response.status).toEqual(StatusCodes.INTERNAL_SERVER_ERROR)
      expect(response.body).toStrictEqual({
        message: 'Server error',
      })
      expect(postService.deletePost).not.toHaveBeenCalled()
    })
  })
})
