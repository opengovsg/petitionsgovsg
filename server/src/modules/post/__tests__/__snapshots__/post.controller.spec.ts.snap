// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostController createPost returns BAD_REQUEST on invalidated request 1`] = `""`;

exports[`PostController createPost returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`PostController createPost returns OK on valid submission 1`] = `
Object {
  "data": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
}
`;

exports[`PostController createPost returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "User not signed in",
}
`;

exports[`PostController getSinglePost retrieves post with specified id 1`] = `
Object {
  "addresseeId": 1,
  "createdAt": "2022-01-01T00:00:00+00:00",
  "email": "John@email.com",
  "fullname": "John Doe",
  "hashedUserSgid": "wxyz",
  "id": "395ac6ec-dea2-4f78-8dc4-1234cafaff87",
  "profile": null,
  "reason": "reason for",
  "references": null,
  "request": "request",
  "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
  "status": "OPEN",
  "summary": null,
  "title": "Basic petition without optional items",
  "updatedAt": "2022-01-01T00:00:00+00:00",
}
`;

exports[`PostController getSinglePost returns BAD_REQUEST on invalidated request 1`] = `""`;

exports[`PostController getSinglePost returns FORBIDDEN on unauthorised user 1`] = `
Object {
  "message": "User does not have permission to access this post",
}
`;

exports[`PostController getSinglePost returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`PostController listPosts retrieves all post 1`] = `
Array [
  Object {
    "addresseeId": 1,
    "createdAt": "2022-01-01T00:00:00+00:00",
    "email": "John@email.com",
    "fullname": "John Doe",
    "hashedUserSgid": "$2b$10$72TKINiR4.ZMaJZmuN5tW.uCuGZcA0TzLUo4qly55GyDndAYG8BSW",
    "id": "af7b6aa9-7c6e-4ea5-b536-4e55bc506847",
    "profile": null,
    "reason": "reason for",
    "references": null,
    "request": "request",
    "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tW.",
    "status": "OPEN",
    "summary": null,
    "title": "Basic petition without optional items",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
  Object {
    "addresseeId": 1,
    "createdAt": "2022-01-01T00:00:00+00:00",
    "email": "John@email.com",
    "fullname": "John Doe",
    "hashedUserSgid": "wxyz",
    "id": "aaaaaaaa-dea2-4f78-8dc4-5542cafaff87",
    "profile": "OGP",
    "reason": "reason",
    "references": "wikipedia.com",
    "request": "request",
    "salt": "$2b$10$72TKINiR4.ZMaJZmuN5tX.",
    "status": "OPEN",
    "summary": "petition summary",
    "title": "Petition with summary, references, profile",
    "updatedAt": "2022-01-01T00:00:00+00:00",
  },
]
`;

exports[`PostController listPosts returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server Error",
}
`;

exports[`PostController publishPost returns FORBIDDEN on post not meeting minimum endorser count 1`] = `
Object {
  "message": "You do not have permission to publish this post.",
}
`;

exports[`PostController publishPost returns FORBIDDEN on user is not post owner 1`] = `
Object {
  "message": "You do not have permission to publish this post.",
}
`;

exports[`PostController publishPost returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Server error",
}
`;

exports[`PostController publishPost returns OK on valid submission 1`] = `
Object {
  "message": "OK",
}
`;

exports[`PostController publishPost returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "You must be logged in to publish posts.",
}
`;

exports[`PostController updatePost returns FORBIDDEN on posts that have already been signed 1`] = `
Object {
  "message": "You do not have permission to update this post.",
}
`;

exports[`PostController updatePost returns INTERNAL_SERVER_ERROR on bad service 1`] = `
Object {
  "message": "Petition failed to update",
}
`;

exports[`PostController updatePost returns INTERNAL_SERVER_ERROR on failed update 1`] = `
Object {
  "message": "Petition failed to update",
}
`;

exports[`PostController updatePost returns OK on valid submission 1`] = `
Object {
  "message": "Petition updated",
}
`;

exports[`PostController updatePost returns UNAUTHORIZED on no user 1`] = `
Object {
  "message": "You must be logged in to update posts.",
}
`;
